app/src/editor/shared/components/RenameInput/index.jsx

```diff
@@ -0,0 +1,48 @@
+// @flow
+
+import React, { useCallback, useRef } from 'react'
+import * as R from 'reactstrap'
+import cx from 'classnames'
+
+import { type Ref } from '$shared/flowtype/common-types'
+import TextInput from '../TextInput'
+import styles from './renameInput.pcss'
+
+type InnerRef = Ref<HTMLTextAreaElement | HTMLInputElement>
+
+type Props = {
+    className?: string,
+    inputClassName?: string,
+    innerRef?: InnerRef,
+}
+
+const PADDING: number = 2
+
+const RenameInput = ({ className, inputClassName, innerRef, ...props }: Props) => {
+    const el: InnerRef = useRef(null)
+    const ref = innerRef || el
+    const onDoubleClick = useCallback(() => {
+        if (ref.current) {
+            ref.current.focus()
+        }
+    }, [ref])
+
+    return (
+        <div
+            className={cx(styles.root, className)}
+            onDoubleClick={onDoubleClick}
+        >
+            <TextInput {...props} innerRef={ref}>
+                {(props) => (
+                    <R.Input
+                        {...props}
+                        className={cx(styles.input, inputClassName)}
```

Comment:
There was `key={hasFocus}` here but it didn't work properly with the "Rename" canvas function (toolbar menu) because of a race condition, kinda (`.focus()` fires first, then the input gets rerendered cause `key` changed).

@timoxley can you explain why do we need that `key={hasFocus}`? Btw, I left it in few places.

Reply:
This is a good question. The intent is actually to flush the browser undo history of the form control on blur. If it doesn't reset the element, then the browser's undo/redo will persist, and this interferes terribly with the app undo/redo.

Reply:
Ha, that explains it pretty well! I'll look into it a bit deeper then.